### PR TITLE
PLSSEM

### PR DESCRIPTION
**Pull request by LSLindeloo**
This analysis still needs some inprovements, especially creating a specialized text area for cSEM

### CODE REVIEW

**Comment by JorisGoosen** [comment]
the new csem model component has been merged

**Comment by LSLindeloo** [comment]
I think the PLS Sem analysis is quite solid now. I decided to remove the nonlinear model options, since many functions in the csem package are not suitable for non linear models. I could still add a test for multiple group differences, predictions of indicator scores of endogenous constructs and prediction metrics, and maybe some other small functions. What do you think of it so far? Do you have any suggestions?

**Review by Kucharssim — CHANGES_REQUESTED** [review]
This looks really impressive! I did not find any errors so far, looks quite robust.

I left a couple of minor code suggestions

Please rebase before changing the code and make sure the tests pass

File: R/sem.R [hunk 1]
```diff
@@ -167,7 +167,7 @@ checkLavaanModel <- function(model, availableVars) {
 
   # Check variable names
   if (!missing(availableVars)) {
-    latents <- unique(parsed[parsed$op == "=~",]$lhs)
+    latents <- unique(parsed[parsed$op == "=~"]$lhs)
```
Comment (Kucharssim) [inline R/sem.R:170]: `parsed` is a data.frame, right? Doesn't this change break SEM then? 
```suggestion
    latents <- unique(parsed[parsed$op == "=~",]$lhs)
```

File: inst/qml/PLSSEM.qml [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+		  	CheckBox { name: "observedIndicatorCorrelation";	label: qsTr("Observed indicator correlations")	}
+			CheckBox { name: "impliedIndicatorCorrelation";		label: qsTr("Implied indicator correlations")	}
+			CheckBox { name: "observedConstructCorrelation"; 	label: qsTr("Observed construct correlations")	}
+			CheckBox { name: "impliedConstructCorrelation"; 	label: qsTr("Implied construct correlations")	}
+		}
+	}
+
+	Section
+	{
+		title: "Prediction"
```
Comment (Kucharssim) [inline inst/qml/PLSSEM.qml:217]:
```suggestion
		title: qsTr("Prediction")
```

File: R/plssem.R [hunk 3]
```diff
@@ -0,0 +1,1882 @@
...
+
+.plsSemPrepOpts <- function(options) {
+  #backwards compatability after changes to bouncontrollavaantextarea.cpp
+  fixModel <- function(model) {
+    newModel <- c(model[1], model[[2]])
+    names(newModel)[names(newModel) == "model"] <- "syntax"
+    return(newModel)
+    }
+
+  options[["models"]] <- lapply(options[["models"]], fixModel)
```
[the remark this answers is not in the document]
Comment (Kucharssim) [reply R/plssem.R:52]: Still think you don't need this ;)

File: R/plssem.R [hunk 4]
```diff
@@ -0,0 +1,1888 @@
...
+  }
+  else if (options[["benchmark"]] == "none") {
+    prediction <- try(cSEM::predict(fit, .handle_inadmissibles = "ignore", .cv_folds = options[["kFolds"]], .r = options[["repetitions"]]))
+  } else {
+    prediction <- try(cSEM::predict(fit, .handle_inadmissibles = "ignore", .benchmark = benchmarks, .cv_folds = options[["kFolds"]], .r = options[["repetitions"]]))
+  }
+  if (isTryError(prediction)) {
+    err <- .extractErrorMessage(prediction)
+    if(grepl("attempt to set 'colnames'", err))
+      err <- "There are not enough observations for each k-fold, try setting 'cross-validation k-folds' to a lower number"
```
Comment (Kucharssim) [inline R/plssem.R:1054]:
```suggestion
      err <- gettext("There are not enough observations for each k-fold, try setting 'cross-validation k-folds' to a lower number")
```

File: R/plssem.R [hunk 5]
```diff
@@ -0,0 +1,1888 @@
...
+    if(grepl("the condition has length > 1", err))
+      err <- "Are all indicator variables set to 'scale'?"
```
Comment (Kucharssim) [inline R/plssem.R:1056]:
```suggestion
      err <- gettext("Are all indicator variables set to 'scale'?")
```

File: R/plssem.R [hunk 6]
```diff
@@ -0,0 +1,1882 @@
...
+      tabr2[["__grp__"]] <- as.list(r2df[["grpname__"]])
+      tabr2[["__var__"]] <- as.list(r2df[["varname__"]])
+      for (i in seq_along(r2li)) tabr2[[paste0("rsq_", i)]] <- r2df[[i + 2]]
+      for (i in seq_along(adjR2li)) tabr2[[paste0("adjustedRsq_", i)]] <- adjR2df[[i + 2]]
+    }
+  }
+}
+
+# Mardias Coefficients table
+.plsSemMardiasCoefficient <- function(modelContainer, dataset, options, ready) {
```
[the remark this answers is not in the document]
Comment (Kucharssim) [reply R/plssem.R:1421]: This still needs to be removed - just use `.semMardiasCoefficient()`

File: R/plssem.R [hunk 7]
```diff
@@ -0,0 +1,1888 @@
...
+  } else {
+    corcont <- createJaspContainer(name, initCollapsed = TRUE)
+  }
+
+  if (options[["group"]] == "") {
+
+    # without groups, these are tables
+
+    if (options[["observedIndicatorCorrelation"]]) {
+      oictab <- createJaspTable("Observed indicator correlation matrix")
```
Comment (Kucharssim) [inline R/plssem.R:1681]:
```suggestion
      oictab <- createJaspTable(gettext("Observed indicator correlation matrix"))
```
the `gettext`s are missing from other tables in this function as well

Comment (Kucharssim) [reply R/plssem.R:1681]: also optionally, you could use `%setOrRetrieve%` (see `` ?jaspBase::`%setOrRetrieve%` ``) for generating the tables to make the code a bit more succinct

**Review by Kucharssim — CHANGES_REQUESTED** [review]
a couple of left over suggestions

**Comment by Kucharssim** [comment]
@LSLindeloo have you committed anything since my review? I see you requested another review but there are no changes...

**Comment by LSLindeloo** [comment]
@Kucharssim Sorry, I was a bit too enthusiastic :P! I will commit trhe changes in a minute!

**Comment by LSLindeloo** [comment]
@Kucharssim The only thing is that, when I remove the `fixModel` function, the analysis does not work, but I don't know why. Do you have any idea?

File: R/plssem.R (hunk 3 above)
[the remark this answers is not in the document]
Comment (Kucharssim) [reply R/plssem.R:52]: > @Kucharssim The only thing is that, when I remove the fixModel function, the analysis does not work, but I don't know why. Do you have any idea?

Ok, well if it doesn't work then perhaps I was wrong and you do need this. You can then leave it as is for now ;)

**Review by Kucharssim — APPROVED** [review]

**Pull request merged**